README.md
# qram
Cram arbitrarily large data into, e.g., multiple streaming QR-codes

## Table of Contents

- [Background](#background)
- [Install](#install)
- [Usage](#usage)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

The primary purpose of this library is to allow arbitrarily large data to
be transmitted using QR codes. It uses [LT Codes][] to pack blocks of data
into packets for efficient delivery via a lossy medium (such as QR codes). The
means of delivery is decoupled from this library so that the data can be
transmitted using any means the programmer is capable to employ. However,
from here forward the examples and documentation focus on delivering the
data via QR codes.

QR-codes can be used to represent relatively small data (e.g., ~7k chars
or ~3k binary data). This library enables arbitrarily large data to be
transferred from one device to another using QR-codes. The term "arbitrarily"
is used loosely here as there will always be various other physical limitations
(e.g., RAM, time, etc.). Extremely large data can be broken into chunks before
being passed to this library, but, not only will transfer time will always be
a limiting factor, constructing a viable transfer system where there is little
to no feedback from the decoder would be challenging.

This README demonstrates how to setup an encoder and decoder for a stream of
QR-codes that can be displayed as a video to a QR-code scanner. The QR-codes
will be efficiently repeated until all of the data has been successfully read
by the scanner.

This project reuses some of the ideas from a similar project written
in Go, [TXQR][].

## Install

- Node.js 8.6+ required.

To install locally (for usage):

```
npm install qram
```

To install locally (for development):

```
git clone https://github.com/digitalbazaar/qram.git
cd qram
npm install
```

## Usage

### Transmitting data

```js
import {Encoder} from 'qram';
// user selected qr-code generator
import QRCode from 'qrcode';
// for converting binary data to text
import * as base64url from 'base64url-universal';

// some data to encode (arbitrarily large)
const data = new Uint8Array([1, 2, 3]);

// create encoder that will produce packets of data for decoder(s)
const encoder = new Encoder({data});

// get a timer for managing frame rate
// TODO: add option to progressively reduce fps after internally calculated
// expected transfer interval
const timer = encoder.createTimer({fps: 30});

// get the stream of packets to efficiently deliver the data to decoder(s)
// stream will indefinitely generate packets to be decoded; stop reading
// from the stream once the decoder has received all of the data
const stream = await encoder.createReadableStream();

// create a function to display the packet as a qr-code
const canvas = document.querySelector('canvas');
const display = ({packet}) =>
  QRCode.toCanvas(canvas, base64url.encode(packet.data));

// keep reading and displaying the packets as qr-code images until the decoder
// has received the data
const reader = stream.getReader();
timer.start();
while(true) {
  // read the next packet
  const {value: packet, done} = await reader.read();
  if(done) {
    break;
  }

  // display the packet as a qr-code for scanning
  await display({packet});

  // manage your frame rate
  // Note: `timer` internally uses `requestAnimationFrame`, if available, to
  // prevent the promise returned from `nextFrame` from resolving until
  // `requestAnimationFrame` runs, preventing changes while the user is
  // not viewing the appropriate window/tab and preventing changes that
  // are faster than the browser itself can render
  await timer.nextFrame();
}

// ... somewhere out-of-band cancel the stream once the decoder has the data
stream.cancel();
```

### Receiving Data

Pick a qr-code reader engine (e.g., [jsQR][]) or use the
[Shape Detection API][]). Wrap your engine of choice in a simple driver API
that takes an image and returns the encoded data as a JavaScript string or
as a Uint8Array.

```js
import {Decoder} from 'qram';
// user selected reader
import jsQR from 'jsqr';
// for converting text to binary data
import * as base64url from 'base64url-universal';

const decoder = new Decoder();

// get a video element to read images of qr-codes from
const video = document.querySelector('video');

// use `requestAnimationFrame` so that scanning will not happen unless the
// user has focused the window/tab displaying the qr-code stream
requestAnimationFrame(function enqueue() {
  // use qram helper to get image data
  const imageData = qram.getImageData(video);
  // use qr-code reader of choice to get Uint8Array or Uint8ClampedArray
  // representing the packet
  const {data: text} = jsQr(imageData.data, imageData.width, imageData.height);
  // enqueue the packet data for decoding, ignoring any errors
  // and rescheduling until done or aborted
  decoder.enqueue(base64url.decode(text)
    .then(progress => {
      // show progress, e.g. `progress.receivedBlocks / progress.totalBlocks`,
      // to user somehow ...

      if(!progress.done) {
        // not done yet, schedule to get another packet
        requestAnimationFrame(enqueue);
      }
    })
    .catch(e => e.name === 'AbortError' ?
      null : requestAnimationFrame(enqueue));
});

try {
  // result found
  const {data} = await decoder.decode();
} catch(e) {
  // failure to decode
  console.error(e);
}

// ... somewhere out-of-band you can cancel if desired
decoder.cancel();
```

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

[New BSD License (3-clause)](LICENSE) © Digital Bazaar

[jsQR]: https://github.com/cozmo/jsQR
[Shape Detection API]: https://wicg.github.io/shape-detection-api/
[TXQR]: https://github.com/divan/txqr
[LT Codes]: https://en.wikipedia.org/wiki/Luby_transform_code


CHANGELOG.md
# qram ChangeLog

## 0.3.9 - 2019-10-03

### Fixed
- Fix validation bug.

## 0.3.8 - 2019-10-03

### Added
- Add type validation to `Packet.create`.

## 0.3.7 - 2019-09-04

### Fixed
- Handle case where decoding is canceled during packet parsing.

## 0.3.6 - 2019-09-04

### Changed
- Provide built in canvas cache for computing image data.

## 0.3.5 - 2019-09-04

### Added
- Add `setRate` API to Timer.

## 0.3.4 - 2019-08-27

### Added
- Expose blocks map in progress event.

## 0.3.3 - 2019-08-26

### Changed
- Decrease minimum header size.

### Added
- Add option to limit max blocks per header.
- Add `Encoder.getMaxPacketSize`.

## 0.3.2 - 2019-08-23

### Fixed
- Fix webpack alias.

## 0.3.1 - 2019-08-23

### Fixed
- Fix incorrect index when decoding.

## 0.3.0 - 2019-08-23

### Changed
- **BREAKING** Include full data hash in packet header.
- Remove isomorphic-webcrypto dependency.

## 0.2.1 - 2019-08-22

### Fixed
- Add webpack build.

## 0.2.0 - 2019-08-22

### Added
- Add initial implementation.

## 0.1.0 - 2019-08-14

### Added
- Add core files.

- See git history for changes previous to this release.


demo/index.js
/*!
 * Copyright (c) 2019 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const {
  QRCode,
  'base64url-universal': base64url,
  jsQR,
  qram: {Decoder, Encoder, getImageData}
} = window;

document.addEventListener('DOMContentLoaded', () => {
  _on('present', 'click', present);
  _on('receive', 'click', receive);
  _on('camera', 'click', toggleCamera);
  _clearProgress();
  _hide('video');
});

const state = {
  decoder: null,
  enableCamera: false,
  runEncoder: false,
  size: 1024,
};

async function toggleCamera() {
  const video = document.getElementById('video');

  if(state.enableCamera) {
    console.log('Camera turned off');
    state.enableCamera = false;
    video.srcObject = null;
    return;
  }

  if(state.runEncoder) {
    // turn off presentation
    present();
  }

  console.log('Camera turned on');
  _hide('canvas');
  _hide('progress');
  _show('video');
  _show('progress');

  const constraints = {
    audio: false,
    video: {
      width: {min: 500},
      height: {min: 500},
      facingMode: 'environment'
    }
  };

  try {
    const stream = await navigator.mediaDevices.getUserMedia(constraints);
    video.srcObject = stream;
  } catch(e) {
    console.error('Failed to setup camera', e);
  }

  state.enableCamera = true;
}

async function present() {
  if(state.runEncoder) {
    console.log('Presentation stopped');
    state.runEncoder = false;
    return;
  }

  // turn off camera
  if(state.enableCamera) {
    toggleCamera();
  }

  _hide('video');
  _hide('progress');
  _show('canvas');
  _show('presenting');

  const size = parseInt(document.getElementById('size').value, 10) || 1024;
  const blockSize = parseInt(
    document.getElementById('block-size').value, 10) || 400;
  let fps = document.getElementById('fps').value;
  const resistance = document.getElementById('resistance').value;

  if(fps !== 'auto') {
    fps = parseInt(fps, 10) || 15;
  } else {
    // rough decent estimate: do `blockCount` frames per second
    fps = Math.min(30, Math.ceil(size / blockSize));
  }

  let sizeMsg;
  if(size < 1024) {
    sizeMsg = `${size} bytes`;
  } else {
    sizeMsg = `${Math.floor(size / 1024)} kiB`;
  }
  const presentMsg =
    `Presenting ${sizeMsg} @ ${fps} frames/second, block size is ` +
    `${blockSize} bytes...`;
  document.getElementById('presenting').innerHTML = presentMsg;
  console.log(presentMsg);

  state.runEncoder = true;

  // generate fake data for presentation
  const data = new Uint8Array(size);
  crypto.getRandomValues(data);

  let version;
  const maxBlocksPerPacket = 50;
  // const maxPacketSize = Encoder.getMaxPacketSize({
  //   size: data.length,
  //   blockSize,
  //   maxBlocksPerPacket
  // });
  // console.log('maxPacketSize', maxPacketSize);

  if(blockSize <= 10) {
    if(resistance === 'H') {
      version = 16;
    } else {
      version = 14;
    }
  } else if(blockSize <= 50) {
    if(resistance === 'H') {
      version = 18;
    } else {
      version = 16;
    }
  } else if(blockSize <= 100) {
    if(resistance === 'H') {
      version = 19;
    } else {
      version = 17;
    }
  } else if(blockSize <= 200) {
    if(resistance === 'H') {
      version = 22;
    } else {
      version = 19;
    }
  } else if(blockSize <= 300) {
    if(resistance === 'H') {
      version = 25;
    } else {
      version = 22;
    }
  } else if(blockSize <= 400) {
    if(resistance === 'H') {
      version = 29;
    } else {
      version = 25;
    }
  }

  const encoder = new Encoder({data, blockSize, maxBlocksPerPacket});
  const timer = encoder.createTimer({fps});
  const canvas = document.getElementById('canvas');
  const stream = await encoder.createReadableStream();
  const reader = stream.getReader();
  timer.start();
  while(state.runEncoder) {
    const {value: packet} = await reader.read();
    const text = base64url.encode(packet.data);
    await QRCode.toCanvas(canvas, text, {
      version,
      //mode: 'alphanumeric',
      errorCorrectionLevel: resistance
    });
    await timer.nextFrame();
  }
}

async function receive() {
  if(state.decoder) {
    console.log('Receive canceled');
    state.decoder.cancel();
    state.decoder = null;
    return;
  }

  _clearProgress();
  _hide('presenting');
  _show('progress');

  let source;
  if(state.enableCamera) {
    console.log('Decoding from camera...');
    // get a video element to read images of qr-codes from
    source = document.getElementById('video');
  } else if(state.runEncoder) {
    console.log('Decoding from canvas directly...');
    // get canvas element to read images of qr-codes from
    source = document.getElementById('canvas');
  } else {
    console.error('Receive aborted, not using camera or presenting locally.');
    return;
  }

  console.log('Scanning...');
  const decoder = state.decoder = new Decoder();

  // use `requestAnimationFrame` so that scanning will not happen unless the
  // user has focused the window/tab displaying the qr-code stream
  requestAnimationFrame(() => setTimeout(enqueue, 0));
  let fails = 0;

  function enqueue() {
    // use qram helper to get image data
    const imageData = getImageData({source});

    const x = performance.now();

    // use qr-code reader of choice to get Uint8Array or Uint8ClampedArray
    // representing the packet
    // const result = jsQR(imageData.data, imageData.width, imageData.height, {
    //   inversionAttempts: 'dontInvert'
    // });

    const idd = imageData.data;
    const image = window.ZXing._resize(imageData.width, imageData.height);
    for(let i = 0, j = 0; i < idd.length; i += 4, j++) {
      window.ZXing.HEAPU8[image + j] = idd[i];
    }
    const err = window.ZXing._decode_qr_multi();
    let result;
    if(err === 0) {
      result = window.ZXing._decode_results[0];
    }

    if(!result) {
      fails++;
      // no QR code found, try again on the next frame
      return requestAnimationFrame(() => setTimeout(enqueue, 0));
    }

    // enqueue the packet data for decoding, ignoring any non-cancel errors
    // and rescheduling until done or aborted
    //const {data: text} = result;
    const text = String.fromCharCode.apply(null, result[0]);
    const y = performance.now();
    console.log('scan time:', (y - x), ' ms');
    const data = base64url.decode(text);
    console.log(`Scanned ${data.length} bytes, parsing...`);
    decoder.enqueue(data)
      .then(progress => {
        if(!progress.done) {
          _updateProgress({progress});
          setTimeout(enqueue, 0);
        } else {
          console.log('total failed scans', fails);
        }
      })
      .catch(e => {
        if(e.name === 'AbortError') {
          return;
        }
        console.error(e);
        setTimeout(enqueue, 0);
      });
  }

  try {
    // result found
    const start = Date.now();
    const progress = await decoder.decode();
    _updateProgress({progress});
    const time = ((Date.now() - start) / 1000).toFixed(3);
    const {data} = progress;
    // console.log('decoded data', data);
    _finish({data, time});
  } catch(e) {
    // failure to decode
    console.error(e);
  }

  state.runEncoder = false;
  state.decoder = null;
}

function _updateProgress({progress}) {
  console.log('Progress', progress);
  const {
    blocks,
    receivedPackets,
    receivedBlocks,
    totalBlocks
  } = progress;
  console.log(`Decoded ${receivedBlocks}/${totalBlocks} blocks`);
  const packetsElement = document.getElementById('packets');
  packetsElement.innerHTML = `Received ${receivedPackets} packets`;
  const blocksElement = document.getElementById('blocks');
  blocksElement.innerHTML = `Decoded ${receivedBlocks}/${totalBlocks} blocks`;
  const blocksMapElement = document.getElementById('blocksmap');
  let blocksMapHTML = '';
  // block width without margin rounded down
  const blockWidth =
    Math.floor((500 - ((totalBlocks - 1) * 1/*px*/)) / totalBlocks);
  for(let i = 0; i < totalBlocks; ++i) {
    const cl = blocks.has(i) ? 'found' : 'missing';
    blocksMapHTML +=
      `<span class="${cl}" style="width: ${blockWidth}px">&nbsp;</span>\n`;
  }
  blocksMapElement.innerHTML = blocksMapHTML;
}

function _finish({data, time}) {
  const size = (data.length / 1024).toFixed(3);
  const msg = `Decoded ${size} KiB in time ${time} seconds`;
  console.log(msg);
  const element = document.getElementById('finish');
  element.innerHTML = `Decoded ${size} KiB in time ${time} seconds`;
}

function _clearProgress() {
  const packets = document.getElementById('packets');
  packets.innerHTML = 'No packets received yet';
  const blocks = document.getElementById('blocks');
  blocks.innerHTML = 'No blocks decoded yet';
  const finish = document.getElementById('finish');
  finish.innerHTML = '';
}

function _on(id, event, listener) {
  const element = document.getElementById(id);
  element.addEventListener(event, listener);
}

function _show(id) {
  document.getElementById(id).style.display = 'block';
}

function _hide(id) {
  document.getElementById(id).style.display = 'none';
}
